Sources/AWSSDKSwiftCore/HTTPClient.swift

```diff
@@ -4,190 +4,276 @@
+        var bootstrap = NIOTSConnectionBootstrap(group: self.eventLoopGroup)
+            .connectTimeout(TimeAmount.seconds(5))
+            .channelOption(ChannelOptions.socket(IPPROTO_TCP, TCP_NODELAY), value: 1)
+        
+        if port == 443 {
+            bootstrap = bootstrap.tlsOptions(NWProtocolTLS.Options())
+        }
+        
+        bootstrap.channelInitializer { channel in
+                return channel.pipeline.addHTTPClientHandlers()
+                    .flatMap {
+                        let handlers : [ChannelHandler] = [
+                            HTTPClientRequestSerializer(hostname: headerHostname),
+                            HTTPClientResponseHandler(promise: response)
+                        ]
+                        return channel.pipeline.addHandlers(handlers)
+                }
+            }
+            .connect(host: hostname, port: port)
+            .flatMap { channel -> EventLoopFuture<Void> in
+                return channel.writeAndFlush(request)
+            }
+            .whenFailure { error in
+                response.fail(error)
+        }
+    }
+    #endif
+    
+    /// send request to HTTP client, return a future holding the Response
+    public func connect(_ request: Request) -> EventLoopFuture<Response> {
+        // extract details from request URL
+        guard let url = URL(string:request.head.uri) else { return eventLoopGroup.next().makeFailedFuture(HTTPClient.HTTPError.malformedURL(url: request.head.uri)) }
+        guard let scheme = url.scheme else { return eventLoopGroup.next().makeFailedFuture(HTTPClient.HTTPError.malformedURL(url: request.head.uri)) }
+        guard let hostname = url.host else { return eventLoopGroup.next().makeFailedFuture(HTTPClient.HTTPError.malformedURL(url: request.head.uri)) }
+
+        let port : Int
+        let headerHostname : String
+        if url.port != nil {
+            port = url.port!
+            headerHostname = "\(hostname):\(port)"
+        } else {
+            let isSecure = (scheme == "https")
+            port = isSecure ? 443 : 80
+            headerHostname = hostname
+        }
+
+
+        let response: EventLoopPromise<Response> = self.eventLoopGroup.next().makePromise()
+
+        #if canImport(Network)
+            if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *) {
+                tsConnectionBootstrap(hostname: hostname, port: port, headerHostname: headerHostname, request: request, response: response)
+            } else {
+                clientBootstrap(hostname: hostname, port: port, headerHostname: headerHostname, request: request, response: response)
+            }
+        #else
+            clientBootstrap(hostname: hostname, port: port, headerHostname: headerHostname, request: request, response: response)
+        #endif
+        
         return response.futureResult
     }
 
+   /// shutdown the event group
     public func close(_ callback: @escaping (Error?) -> Void) {
```

Comment:
Maybe we should also [create a `deinit` where we check to make sure that `close` was called](https://github.com/swift-server/async-http-client/blob/master/Sources/AsyncHTTPClient/HTTPClient.swift#L68) ?